addons/udes_stock/models/product_product.py

```diff
@@ -63,3 +63,28 @@ def unlink(self):
         """Override superclass to prevent deletion."""
         raise ValidationError(_("Products may not be deleted. Please archive them instead."))
 
+    def get_quant_counts(self):
+        warehouse = self.env.ref("stock.warehouse0")
+        Quant = self.env["stock.quant"]
+        quants = Quant.search_count(
+            [
+                ("product_id", "in", self.ids),
+                ("location_id", "child_of", warehouse.view_location_id.id),
+            ]
+        )
+        return quants
+
+    def has_goods_in_transit_or_stock(self):
+        Picking = self.env["stock.picking"]
+        picking_type = self.env.ref("stock.picking_type_in")
+        stock_pickings = Picking.search(
+            [
+                ("move_lines.product_id", "in", self.ids),
+                ("state", "=", "assigned"),
+                ("picking_type_id", "=", picking_type.id),
+            ]
+        )
+
+        if self.get_quant_counts() > 0 or bool(stock_pickings):
```

Comment:
I would swap the conditions so the quant search is not triggered if there are pickings. Also we can directly return this condition.